runtime/balrt.h

```diff
@@ -247,11 +253,77 @@ static READONLY inline int64_t taggedToInt(TaggedPtr p) {
     }
 }
 
+static READONLY inline int64_t taggedPrimitiveCompare(TaggedPtr lhs, TaggedPtr rhs, int64_t(*comparator)(TaggedPtr, TaggedPtr)) {
+    bool isLhsNull = lhs == NULL;
+    bool isRhsNull = rhs == NULL;
+    if (isLhsNull && isRhsNull) {
+        return EQ;
+    }
+    else if (isLhsNull || isRhsNull) {
+        return UN;
+    }
+    return (*comparator)(lhs, rhs);
+}
+
+static READONLY inline int64_t taggedIntComparator(TaggedPtr lhs, TaggedPtr rhs) {
+    int64_t lhsVal = taggedToInt(lhs);
+    int64_t rhsVal = taggedToInt(rhs);
+    if (lhsVal == rhsVal) {
+        return EQ;
+    }
+    if (lhsVal < rhsVal) {
+        return LT;
+    }
+    if (lhsVal > rhsVal) {
+        return GT;
+    }
+    return UN;
+}
+
+static READONLY inline int64_t taggedIntCompare(TaggedPtr lhs, TaggedPtr rhs) {
+    int64_t(*comparator)(TaggedPtr, TaggedPtr) = &taggedIntComparator;
+    return taggedPrimitiveCompare(lhs, rhs, comparator);
```

Comment:
Get rid of the first line and use `taggedIntComparator` instead of comparator.

Does clang inline all this? If not, then don't use function pointers.

Reply:
Fixed